### PR TITLE
dzhyv - Recently viewed products displayed from latest viewed product to first

### PR DESCRIPTION
**Pull request by dzhyv**

### CODE REVIEW

**Review by Taras-Kr — APPROVED** [review]

**Review by anastasia8755 — APPROVED** [review]

**Review by NyxLex — APPROVED** [review]

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 1]
```diff
@@ -33,11 +36,20 @@ public boolean isOpened() {
...
         return $x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", itemNumber)).getText();
     }
 
+    @Step("HomePage: get recently viewed products names list from first to {intNumber} Recently Viewed Products list")
+    public List<String> getRecentlyViewedProductsNames(int itemNumber){
+        List<String> names = new ArrayList<>();
+        for (int i = 1; i <= itemNumber; i++){
+            names.add($x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", i)).getText());
+        }
+        return names;
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:50]: use $x().texts() instead

File: src/test/java/com/ss/ita/rozetka/test/RecentlyViewedProductsTest.java [hunk 2]
```diff
@@ -36,6 +42,39 @@ public void verifyLastViewedProductAddedToTheList() {
...
+        String secondOpenedProductName = productPage
+                .getHeader()
+                .openHomePage()
+                .openProductCategoryPage(COTTAGE_GARDEN_BACKYARD)
+                .openProductTypePage(GARDEN_TECH_CATEGORY)
+                .openProductPage(1)
+                .getName();
+
+        String thirdOpenedProductName = productPage
+                .getHeader()
```
Comment (toleksyn) [inline src/test/java/com/ss/ita/rozetka/test/RecentlyViewedProductsTest.java:56]: re-use line 48

Comment (toleksyn) [reply src/test/java/com/ss/ita/rozetka/test/RecentlyViewedProductsTest.java:56]: same for all

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 3]
```diff
@@ -33,11 +35,17 @@ public boolean isOpened() {
...
     }
 
-    @Step("HomePage: get product number {itemNumber} from Recently Viewed Products list")
+    @Step("HomePage: get product name number {itemNumber} from Recently Viewed Products list")
     public String getRecentlyViewedProductName(int itemNumber) {
         return $x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", itemNumber)).getText();
     }
 
+    @Step("HomePage: get recently viewed products names list from first to {intNumber} Recently Viewed Products list")
+    public List<String> getRecentlyViewedProductsNames(/*int itemNumber*/){
```
Comment (Percifalll) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:44]: no commented code in PR

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 4]
```diff
@@ -33,11 +35,17 @@ public boolean isOpened() {
...
+        $x("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li//a[@class = 'tile__title']").shouldBe(Condition.visible);
+        return  $$x("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li//a[@class = 'tile__title']").texts();
```
Comment (Percifalll) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:46]: move xpath string to local variable

**Review by Percifalll — CHANGES_REQUESTED** [review]
.

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 5]
```diff
@@ -33,11 +35,18 @@ public boolean isOpened() {
...
+    @Step("HomePage: get product name number {itemNumber} from Recently Viewed Products list")
     public String getRecentlyViewedProductName(int itemNumber) {
         return $x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", itemNumber)).getText();
     }
 
+    @Step("HomePage: get recently viewed products names list from first to {intNumber} Recently Viewed Products list")
+    public List<String> getRecentlyViewedProductsNames(){
+        String recentlyViewedProductsXpath = "//section[@class = 'main-goods ng-star-inserted'][1]//ul/li//a[@class = 'tile__title']";
+        $x(recentlyViewedProductsXpath).shouldBe(Condition.visible);
+        return  $$x(recentlyViewedProductsXpath).texts();
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:47]: $$x() has no waits

File: src/test/java/com/ss/ita/rozetka/test/RecentlyViewedProductsTest.java [hunk 6]
```diff
@@ -27,15 +33,48 @@ public void verifyLastViewedProductAddedToTheList() {
...
+                .as("First product name in Recently Opened products should be equal to last viewed product name")
+                .isEqualTo(thirdOpenedProductName);
+
+        assertThat(recentlyViewedProductNames.get(1))
+                .as("Second product name in Recently Opened products should be equal to second viewed product name")
+                .isEqualTo(secondOpenedProductName);
+
+        assertThat(recentlyViewedProductNames.get(2))
+                .as("Third product name in Recently Opened products should be equal to first viewed product name")
+                .isEqualTo(firstOpenedProductName);
```
Comment (toleksyn) [inline src/test/java/com/ss/ita/rozetka/test/RecentlyViewedProductsTest.java:78]: think about soft assert

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Review by Percifalll — APPROVED** [review]

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 7]
```diff
@@ -33,11 +36,16 @@ public boolean isOpened() {
...
 
-    @Step("HomePage: get product number {itemNumber} from Recently Viewed Products list")
+    @Step("HomePage: get product name number {itemNumber} from Recently Viewed Products list")
     public String getRecentlyViewedProductName(int itemNumber) {
         return $x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", itemNumber)).getText();
     }
 
+    @Step("HomePage: get recently viewed products names list from first to {intNumber} Recently Viewed Products list")
+    public List<String> getRecentlyViewedProductNames(){
+        return  $$x("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li//a[@class = 'tile__title']").shouldBe(sizeNotEqual(0)).texts();
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:46]: format

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java [hunk 8]
```diff
@@ -33,11 +36,16 @@ public boolean isOpened() {
...
 
-    @Step("HomePage: get product number {itemNumber} from Recently Viewed Products list")
+    @Step("HomePage: get product name number {itemNumber} from Recently Viewed Products list")
     public String getRecentlyViewedProductName(int itemNumber) {
         return $x(format("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li[%s]//a[@class = 'tile__title']", itemNumber)).getText();
     }
 
+    @Step("HomePage: get recently viewed products names list from first to {intNumber} Recently Viewed Products list")
+    public List<String> getRecentlyViewedProductNames() {
+        return $$x("//section[@class = 'main-goods ng-star-inserted'][1]//ul/li//a[@class = 'tile__title']").shouldBe(sizeNotEqual(0)).texts();
```
Comment (toleksyn) [inline src/main/java/com/ss/ita/rozetka/pageobject/pages/HomePage.java:46]: format

**Review by toleksyn — CHANGES_REQUESTED** [review]

**Pull request merged**